File: java/sdk/sdk-pay/src/main/java/com/novax/sdk/pay/receipt/ReceiptAddOrderRequest.java
package com.novax.sdk.pay.receipt;

import com.novax.sdk.core.request.AbstractApiRequest;
import com.novax.sdk.core.request.HttpMethod;
import com.novax.sdk.core.request.TypeRef;
import com.novax.sdk.pay.model.ReceiptOrderAddressResp;
import lombok.Builder;
import lombok.Getter;

import java.math.BigDecimal;
import java.util.LinkedHashMap;
import java.util.Map;

/** POST /pay/v3/receipt/order/add — create a receipt order. */
@Getter
@Builder
public class ReceiptAddOrderRequest extends AbstractApiRequest<ReceiptOrderAddressResp> {

    private final String protocol;
    private final String currency;
    private final String smartContractAddress;
    private final String companyUserId;
    private final String receiptOrderId;
    private final BigDecimal currencyNumber;
    private final String callBackUrl;

    @Override public HttpMethod method() { return HttpMethod.POST; }
    @Override public String path() { return "/pay/v3/receipt/order/add"; }

    @Override public Object body() {
        Map<String, Object> b = new LinkedHashMap<>();
        b.put("protocol", protocol);
        b.put("currency", currency);
        b.put("smartContractAddress", smartContractAddress);
        b.put("companyUserId", companyUserId);
        b.put("receiptOrderId", receiptOrderId);
        b.put("currencyNumber", currencyNumber);
        b.put("callBackUrl", callBackUrl);
        return b;
    }

    @Override public TypeRef<ReceiptOrderAddressResp> responseType() {
        return TypeRef.of(ReceiptOrderAddressResp.class);
    }
}

File: python/novax_sdk/pay/receipt/receipt_add_order_request.py
from dataclasses import dataclass
from decimal import Decimal
from typing import Any

from novax_sdk.pay.model import ReceiptOrderAddressResp
from novax_sdk.request.api_request import ApiRequest, HttpMethod


@dataclass
class ReceiptAddOrderRequest(ApiRequest):
    """POST /pay/v3/receipt/order/add"""

    protocol: str | None = None
    currency: str | None = None
    smart_contract_address: str | None = None
    company_user_id: str | None = None
    receipt_order_id: str | None = None
    currency_number: Decimal | None = None
    call_back_url: str | None = None

    @property
    def method(self) -> HttpMethod:
        return HttpMethod.POST

    @property
    def path(self) -> str:
        return "/pay/v3/receipt/order/add"

    def body(self) -> dict[str, Any]:
        return {
            "protocol": self.protocol,
            "currency": self.currency,
            "smartContractAddress": self.smart_contract_address,
            "companyUserId": self.company_user_id,
            "receiptOrderId": self.receipt_order_id,
            "currencyNumber": str(self.currency_number) if self.currency_number else None,
            "callBackUrl": self.call_back_url,
        }

    @property
    def response_type(self) -> type:
        return ReceiptOrderAddressResp
